presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.LongArrayBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import io.airlift.slice.SliceOutput;
+
+import java.util.Arrays;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.max;
+
+public abstract class BlockEncodingBuffers
+{
+    private static final int BITS_IN_BYTE = 8;
+    private static final int INITIAL_POSITION_COUNT = 1024;
+
+    protected int[] positions;
+    protected int[] mappedPositions;
+    protected int positionCount;
+    protected boolean isPositionsMapped;
+
+    protected int positionsOffset;  // each batch we copy the values of rows from positions[positionsOffset] to positions[positionsOffset + batchSize]
+    protected int batchSize;
+    protected int bufferedPositionCount;
+
+    protected int initialBufferSize;
+
+    protected byte[] nullsBuffer;
+    protected int nullsBufferIndex;  //The next byte address if new values to be added.
+    protected boolean[] remainingNullsFromLastBatch = new boolean[BITS_IN_BYTE];
+    protected int remainingNullsCount;
+    protected boolean nullsBufferContainsNull;
+
+    protected Block decodedBlock;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(PartitionedOutputOperator.DecodedObjectNode decodedObjectNode, int initialBufferSize)
+    {
+        Object decodedBlock = decodedObjectNode.getDecodedBlock();
+
+        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
+        if (decodedBlock instanceof DictionaryBlock) {
+            DictionaryBlock dictionaryBlock = (DictionaryBlock) decodedBlock;
+            decodedBlock = dictionaryBlock.getDictionary();
+        }
+        else if (decodedBlock instanceof RunLengthEncodedBlock) {
+            decodedBlock = ((RunLengthEncodedBlock) decodedBlock).getValue();
+        }
+
+        if (decodedBlock instanceof LongArrayBlock) {
+            return new LongArrayBlockEncodingBuffers(initialBufferSize);
+        }
+        else {
+            throw new IllegalArgumentException("Unsupported encoding: " + ((Block) decodedBlock).getEncodingName());
```

Comment:
cast to Block is not safe; better use class name

`throw new IllegalArgumentException("Unsupported encoding: " + decodedBlock.getClass().getSimpleName());`